### PR TITLE
feat(appset): ignoreApplicationDifferences (#9101)

### LINKED ISSUE #9101: ApplicationSet: Add option to preserve manual parameter overrides
# Summary

ApplicationSets automatically override manual Application parameter overrides, e.g. performed with argocd client (argocd app set command). The only option to avoid it seems to be to start the applicationset controller with "create-only" policy but this is not always what one wants. Especially in combination with a pull request generator since the pull request apps will never be deleted again once a pull request has been closed.

# Motivation

I agree that manual parameter overrides should never be used in production but there is a dev use case where we need them:

1. Inside our pull request pipeline we're creating a temporary staging docker repository on the fly and we push our PR docker image into it
2. The application generated by PR generator should consume images from that prev. created staging repository
3. The name of the staging repository including the credentials to access it are not predictable and therefore we use argocd client to override these parameters

Inside our main pipeline basically the same mechanism happens. We create a staging repository and push our docker image into it. Then, we want to do artifact validation by running a bunch of tests and only if they are successful we promote our docker image from staging into our final release repository.

# Proposal

How do you think this should be implemented?

There should be a switch on applicationset level to express that manual parameter overrides should be preserved. If that switch is true and applicationset controller updates an Application object it should merge the parameter overrides instead of just replacing them with whatever is inside the repository.

### PR DESCRIPTION
**Pull request by crenshaw-dev**
Closes #9101

Concerns:
1) this implementation piggy-backs on the more general `ignoreDifferences` feature of the application controller and will suffer from all its current limitations - Blake is trying to improve on that here: https://github.com/argoproj/argo-cd/pull/14602
2) performance - is all the json marshaling/unmarshaling going to cause slowness?

### CODE REVIEW

**Comment by jessebye** [comment]
@crenshaw-dev Anything blocking this? I'd love to help get this shipped if there's anything I can do :smile:

**Comment by crenshaw-dev** [comment]
@jessebye I think more testing would be _good_, but could probably ship without it.

**Review by ishitasequeira — APPROVED** [review]
The changes LGTM!!

File: docs/operator-manual/applicationset.yaml [hunk 1]
```diff
@@ -26,3 +26,13 @@ spec:
     preserveResourcesOnDeletion: false
   # Alpha feature to determine the order in which ApplicationSet applies changes.
   strategy:
+  # This field lets you define fields which should be ignored when applying Application resources. This is helpful if
```
[the remark this answers is not in the document]
Comment (crenshaw-dev) [reply docs/operator-manual/applicationset.yaml:29]:
```suggestion
  # This field lets you define fields which should be ignored when applying Application resources. This is helpful if you
```

**Comment by LS80** [comment]
I have a problem in `v2.9.0-rc1` with every ApplicationSet that includes `ignoreDifferences` in the Application spec template.

```
 "" is invalid: patch: Invalid value
```
Could it be related to this change?

**Comment by YR-ZR0** [comment]
Same here also results in a degraded state in ArgoCD

**Comment by crenshaw-dev** [comment]
That's surprising... I doubt it's related to this PR, but maybe. Can someone open an issue to track the regression?

**Comment by alexelent-carta** [comment]
I just tried testing this and if I click refresh in the argocd ui, my manual overrides are lost.

```
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: foobar
  namespace: argocd
spec:
  goTemplate: true
  goTemplateOptions: ["missingkey=error"]
  # so we can use argocd cli to set params without appset overriding
  ignoreApplicationDifferences:
    - jqPathExpressions:
        - .spec.source.helm.values
        - .spec.source.targetRevision
  generators:
    # matrix "parent" generator
    - matrix:
        generators:
          # cluster generator - all clusters
          - clusters: {}
          - git:
              repoURL: https://github.com/foo/bar.git
              revision: master
              directories:
                - path: helm/*
```


```
$ ARGOCD_SERVER=foobar.com:443 \
        ARGOCD_AUTH_TOKEN=FOOBAR \
        argocd app set foobar -p tag=1.2.3 \
        --grpc-web \
        --revision 1.2.3
```

**Comment by crenshaw-dev** [comment]
`argocd app set foobar -p tag=1.2.3` would set `.spec.source.helm.parameters`, not `.spec.source.helm.values`.

**Comment by alexelent-carta** [comment]
Setting to `spec.source.helm.parameters` does not work as well. I am using multiple sources so I even tried `.spec.sources[] | select(.repoURL | contains("foobar")).helm.parameters` to no avail.

**Comment by crenshaw-dev** [comment]
@alexelent-carta really good find. The fix ended up being more involved than I expected. Can you give it a try? https://github.com/argoproj/argo-cd/pull/15965

Since you're using `sources` instead of `source`, you're going to suffer from a limitation that also applies to the Application's `ignoreDifferences`. When the Kubernetes client encounters a list where _any_ field has changed, it will patch the whole list. In other words, as long as _all_ modified fields are present in your `ingoreApplicationDifferences` config, your changes will not be blown away. But them moment any non-ignored field changes, the whole list will be patched.

We might be able to mitigate that issue by adding a new merge-key field to the `sources` items. But adding that required field would, I think, be a breaking change.

For now you might be best off just ignoring all changes in the whole `sources` field.

**Comment by alexelent-carta** [comment]
> For now you might be best off just ignoring all changes in the whole sources field.

This would solve my use case so I am game for this.

Just tested and it works 🎉 💪 . You are my hero.

**Comment by crenshaw-dev** [comment]
@alexelent-carta awesome, I'll put your use case in a test so it doesn't get missed again. :-)

**Comment by crenshaw-dev** [comment]
The fix PR will [document](https://github.com/argoproj/argo-cd/pull/15965/files#diff-c51866835fc818288e35b99ea6a51bf0e1ae27b9140016fdd0f0d48220c767da) the limitation, and I opened an [issue](https://github.com/argoproj/argo-cd/issues/15975) to track the long-term fix.

**Comment by amine7777** [comment]
We are willing to use ignoreApplicationDifferences but as we re using sources and we don't want to ignore all the elements in sources this is a limitation. Also we set sources to empty list [] and we use templatePatch instead

https://github.com/argoproj/argo-cd/pull/14743#issuecomment-1762216897

**Pull request merged**